### ISSUE TITLE
8.2 release

### ISSUE THREAD

**Issue by peastman** [openmm/openmm#4659]
Now that the HIP platform is in, I think we can start moving forward on the 8.2 release.  There are a couple of open PRs I'd like to see if we can get finished up (#4633 and #4647) before creating the beta.  Does anyone have anything else we should try to get in?

**Comment by peastman** [comment]
#4633 is merged, and we decided not to include #4647 in 8.2.  I'll start work on creating the beta.

**Comment by mikemhenry** [comment]
Do we have a draft of the release notes for 8.2?

**Comment by mattwthompson** [comment]
I tried using the UI to compare between 8.1.2 and the commit used in the most recent feedstock PR: https://github.com/openmm/openmm/compare/8.1.2...0f67a646a93a7fa600ae238b6485c54cd3c553db

That's 82 commits going back almost a year - are those changes from as far back as November 2023 not in any 8.1.x release, or am I comparing against the wrong tag?

**Comment by peastman** [comment]
That's correct, it's been a while.  I haven't written the release notes yet.

**Comment by peastman** [comment]
The beta packages are now up.  You can install them with

```
mamba install -c conda-forge/label/openmm_rc -c conda-forge openmm
```

or if you also want the HIP platform,

```
mamba install -c conda-forge/label/openmm_rc -c conda-forge openmm-hip
```

You don't need to specify both packages.  `openmm-hip` installs `openmm` as well.

Next I'll create the github release and post an announcement on the forum.

This is the time for testing and getting feedback, so please do both!  If all goes well, we'll plan to create the release candidate in about a month.

**Comment by peastman** [comment]
The release notes are at https://github.com/openmm/openmm/releases/tag/8.2.0beta.

**Comment by egallicc** [comment]
I installed `openmm-hip` from `conda-forge/label/openmm_rc` as indicated.

The HIP platform fails with
```
:0:/home/conda/feedstock_root/build_artifacts/hip_1718643748184/work/clr/hipamd/src/hiprtc/hiprtcInternal.hpp:105 : 1162859498287 us: [pid:661320 tid:0x7f61459e7740] Unable to add internal header
Aborted (core dumped)
```

The OpenCL platform fails with:
```
openmm.OpenMMException: Error compiling kernel: fatal error: malformed or corrupted AST file: 'could not find file '/software/emilio/miniforge3/envs/atm8.2.0beta/lib/clang/17/include/opencl-c-base.h<F5><FF><FF><FF><FF><FF><FF><FF><U+000F><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000>' referenced by AST file '/tmp/comgr-1bb20c/include/opencl1.2-c.pch''
1 error generated.
Error: Failed to compile source (from CL or HIP source to LLVM IR).
```
The same behavior persists after installing `clang=17` and `hipcc`. 

This is on  Linux x86_64 Ubuntu 22.04.4 LTS ROCM 6.0.2 with RX 6650 XT + RX 6750 XT cards.

OpenMM 8.1.1 + HIP plugin built from sources works fine on this system.

**Comment by ex-rzr** [comment]
@egallicc 
I'm able to reproduce this crash, on **miniforge+mamba**. But the HIP platform works on **anaconda+conda**!
I don't understand the difference: if I compare `ldd <...>/lib/plugins/libOpenMMHIP.so` for both cases, the outputs look very similar.

If anyone is going to experiment with it, please don't forget to run `/tmp/openmm-hip*`  for removing cached kernels and force recompiling in the new environment.

**Comment by ex-rzr** [comment]
Weird, `libhiprtc.so.6.0.32831-58582f0` in miniforge+mamba and anaconda+conda are different! Though they have exactly the same size.
![изображение](https://github.com/user-attachments/assets/6365b3b4-2a38-4c56-afd7-89bf0e1b2eb1)

mamba modifies paths in shared objects. https://github.com/mamba-org/mamba/blob/main/docs/source/advanced_usage/detailed_operations.rst#the-package-installation-process

The problem is that these paths are in a huge string with builtin headers for hipRTC, and even the new data is padded with zeros, this breaks loading of the headers here: https://github.com/ROCm/clr/blob/rocm-6.2.x/hipamd/src/hiprtc/hiprtcInternal.cpp#L235-L242

As far as I know conda does the same modification to paths, but not in this case.

@peastman Did you test the HIP platform with mamba installation earlier? Or was it conda?

**Comment by egallicc** [comment]
Thank you @ex-rzr. Interesting. I think we are looking at a hardware-related issue. It appears that ROCM 6 [dropped support](https://rocm.docs.amd.com/projects/install-on-linux/en/latest/reference/system-requirements.html) for the RDNA 2 cards I have. OpenMM 8.2 built from sources against ROCM 5 works fine on this same system.

Also, I think the OpenCL issue can be remedied by installing the `rocm-opencl` and/or `rocm-opencl-runtime` packages. OpenCL still fails but an error similar to the HIP platform suggesting a common hardware-related issue.

Anyway, apart from the hardware issue with ROCM 6 packaged in conda-forge, so far my tests confirm that ATMForce() works correctly in OpenMM 8.2 on all platforms, and it is significantly faster than 8.1 as previous tests indicated. Thanks!

**Comment by peastman** [comment]
> @peastman Did you test the HIP platform with mamba installation earlier? Or was it conda?

It's a miniconda installation, but I use mamba for installing packages.  The HIP platform works correctly, but OpenCL fails with the same error @egallicc posted.  (I previously described it at https://github.com/openmm/openmm/issues/4623#issuecomment-2322255161.)

> I think the OpenCL issue can be remedied by installing the rocm-opencl and/or rocm-opencl-runtime packages.

I'll try that and see what happens.

**Comment by peastman** [comment]
Installing `rocm-opencl` and `rocm-opencl-runtime` doesn't affect the OpenCL error.  HIP still works fine though.

> I think we might be looking at a hardware-related issue. It appears that ROCM 6 [dropped support](https://rocm.docs.amd.com/projects/install-on-linux/en/latest/reference/system-requirements.html) for the RDNA 2 cards I have.

I'm using an RDNA 1 card.

**Comment by egallicc** [comment]
Thank you. I can confirm @ex-rzr observation. openmm-hip installed from conda works fine on my system.

**Comment by peastman** [comment]
@egallicc I created a new issue from your post: #4675.  Let's continue the discussion there.

**Comment by peastman** [comment]
I think we may be about ready to build the release candidate.  Is there anything else that still needs to be done first?

**Comment by peastman** [comment]
Conda packages for the release candidate are up.  You can install them with

```
mamba install -c conda-forge/label/openmm_rc -c conda-forge openmm
```

We do *not* have conda packages for the HIP platform.  We've been unable to find any solution for the problems in #4675.  It appears it may simply be impossible to support HIP through conda-forge.  Their packages for the HIP runtime are broken.  Installing them breaks both the HIP and OpenCL platforms.  But I can't find any possible way to install the OpenMM HIP platform through conda-forge without it also installing the broken runtime packages.

Instead, I think we're going to need to accelerate support for installing with pip instead.  Packages for the release candidate are now up on the test server.  You can install them with

```
pip install --pre --index-url https://test.pypi.org/simple/ --extra-index-url https://pypi.org/simple/ openmm
```

That package includes the reference, CPU, and OpenCL platforms.  To get the CUDA and/or HIP platform, specify `openmm[cuda12]` or `openmm[hip6]`.  The latter requires that you've installed HIP by following the instructions at https://rocm.docs.amd.com.

Please try them out and post whether they work for you.  If everything is looking ok, I'll post them to the main PyPI server and invite more people to test them.  Once we're confident of them, I'll update the manual to describe both methods of installing, and the 8.2 release will support both of them.

**Comment by kiyoon** [comment]
@peastman Thank you. I tested this on Mac with numpy 2.0 and can confirm that existing tests passed for my program.

**Comment by mattwthompson** [comment]
Should we be testing against the 8.2.0 or 8.2.0rc1 builds on https://test.pypi.org/project/OpenMM/#history?

**Comment by peastman** [comment]
8.2.0rc1.

**Comment by mattwthompson** [comment]
Are there plans to make x86-based macOS wheels?

PyPI packages are looking good on our end in every other way I've been testing!

**Comment by peastman** [comment]
Thanks!

> Are there plans to make x86-based macOS wheels?

I wasn't planning on it.  Would it be useful?  I was figuring most people were now on ARM.

**Comment by kiyoon** [comment]
@peastman I personally don't need it, but there are certainly some people who are on intel still. I guess if you build with GitHub Actions, it should be as easy as adding one more job. If you look for the future it's not necessary, but making it work for the best compatibility isn't much of a hassle either, in my opinion

**Comment by peastman** [comment]
I'm attempting to build them at https://github.com/peastman/openmm-build-wheels/pull/18.  For some reason it's *only* building for x86, not ARM.  I'm not sure what's going on.

**Comment by kiyoon** [comment]
@peastman I have a feeling that that "include" statement overwrites the previous value?

I used to have it named "platform" and use `${{ matrix.platform.os}}`

**Comment by peastman** [comment]
We ought to get one build for each include.  See https://docs.github.com/en/actions/writing-workflows/workflow-syntax-for-github-actions#jobsjob_idstrategymatrixinclude.

**Comment by j-wags** [comment]
I think (4/10) it's worth keeping builds for Intel Macs if it's not a huge effort. The build matrix starts at Python 3.9, which was released in 2020, while Apple only stopped selling Intel Macs (at that time just the mini) in 2023. So if we're just going by dates, it makes a bit of sense to include them for a while longer.

**Comment by peastman** [comment]
I found a problem that's going to require doing another release candidate.  HipContext has a `createFFT()` method to create a new HipFFT3D object.  This is a remnant of when it was supporting multiple FFT implementations.  So including `HipContext.h` causes it to include `HipFFT3D.h`, which includes `vkFFT.h`, which does not get installed.  The upshot is that it's impossible for another package to provide a HIP implementation of a new feature, because including `HipContext.h` leads to a compilation error.  I'll get a fix in shortly.

**Comment by kiyoon** [comment]
@peastman Thanks for your work!

**Comment by ex-rzr** [comment]
@peastman 

Oh, indeed. I assume `createFFT()` calls should be replaced with `new HipFFT3D()`, correct?

**Comment by peastman** [comment]
Right.  The fix is in #4714.

**Comment by peastman** [comment]
The pip packages for 8.2rc2 are up.  This change doesn't affect the conda packages, because the only changes are in the HIP platform which they don't include.

**Comment by peastman** [comment]
I think we're ready to make the release.  Unless someone objects, I'm going to move forward with it.

**Comment by jchodera** [comment]
No issues have been raised in response to the twitter thread or posts in OpenFF, OpenFE, or OMSF Slacks.

**Comment by peastman** [comment]
I created the github release.  Full release notes are at https://github.com/openmm/openmm/releases/tag/8.2.0.

Pip packages are now up.

Conda packages are building now.

**Comment by peastman** [comment]
Conda packages are now up.  I also updated the online documentation.

**Comment by peastman** [comment]
I created releases and built conda packages for PDBFixer, OpenMM-Setup, and OpenMM-Torch.

I also built a pip package for PDBFixer, but when I tried to upload it I got an error that the name wasn't permitted.  I'm not sure why, since as far as I can tell there is no existing package with that name.  I opened an issue about it.

I also tried to build pip packages for OpenMM-Torch, but it's blocked by a PyTorch bug (https://github.com/pytorch/pytorch/issues/139108).

**Comment by peastman** [comment]
I'm still stuck on trying to upload the pip package for PDBFixer.  As far as I can tell there is no package with that name, but it tells me the name isn't allowed.  I opened an issue about it over a week ago (https://github.com/pypi/support/issues/5093), but there's been no reply.  Does anyone have any idea what I could do to resolve it?

**Comment by mikemhenry** [comment]
Maybe this package is the issue? https://pypi.org/project/black-pdbfixer/

**Comment by mpds** [comment]
Any updates on getting PDBFixer available via pip? Conda's been pushing my patience; a pip version would be great.

**Comment by peastman** [comment]
Nothing unfortunately.  I'd really appreciate help if anyone has experience with this, or contacts in the PyPI organization, or anything like that.  The current situation is that it won't let me post the package.  I don't know why it won't.  I don't know how to determine what's causing the problem.  Since I don't know the cause, I don't know how to fix it.  And the people who could help aren't responding.

**Comment by bdenhollander** [comment]
Did you reach out to the maintainer of [black-pdbfixer](https://pypi.org/project/black-pdbfixer/) regarding removing that package in favour of an official one? There's an email address shown when hovering over Peter Eastman in the left panel.

**Comment by peastman** [comment]
I'm going to close this issue.  8.2 is out, and the other pypi packages are blocked by other issues.
